### PR TITLE
Basic Declaritive Network Configuration in wpa_supplicant Service

### PR DESCRIPTION
**Pull request by globin**
- remove obsolete option networking.WLANInterface, it has been obsolete for years (according to `git blame`)
- refactored to `systemd.services`
- removed `preStart` in the service which created `/etc/wpa_supplicant.conf` if it didn't exist and `userControlled` was turned on, the code is added to the declarative `wpa_supplicant.conf` if used, but does not change existing setups.
- added basic declarative networks specification currently only supporting open wifi networks and simple PSK authentication

This is backward compatible to current configuration by defaulting to using `/etc/wpa_supplicant.conf` if `networking.wireless.networks` is empty.

### CODE REVIEW

**Comment by copumpkin** [comment]
Possibly of interest, except it depends on a primop that never made it in: https://github.com/edolstra/nixpkgs/commit/4c8212069429bf9fb959e00ce8d9345ac7cb7ff0

**Comment by globin** [comment]
Just found an error but don't have any network connection apart from a very flaky one on my phone, will fix later.

**Comment by globin** [comment]
Ah I actually managed to push

**Comment by dezgeg** [comment]
Seems to break some test evaluation: http://hydra.nixos.org/jobset/nixos/trunk-combined#tabs-errors
````
in job ‘tested’:
The unique option `systemd.services.wpa_supplicant.description' is defined multiple times, in `/nix/store/17l7cmx1p5s8yq05jk2cf5dg0mws960w-git-export/nixos/modules/services/networking/wpa_supplicant.nix' and `/nix/store/17l7cmx1p5s8yq05jk2cf5dg0mws960w-git-export/nixos/modules/system/upstart/upstart.nix'.
````

**Comment by peti** [comment]
Okay, this is bad. It would be great to get a fix for that issue soon, otherwise we'll have to revert.

**Comment by domenkozar** [comment]
I'm looking into it.

File: nixos/doc/manual/configuration/wireless.xml [hunk 1]
```diff
@@ -18,8 +18,18 @@ NixOS will start wpa_supplicant for you if you enable this setting:
 networking.wireless.enable = true;
 </programlisting>
 
-NixOS currently does not generate wpa_supplicant's
-configuration file, <literal>/etc/wpa_supplicant.conf</literal>. You should edit this file
+NixOS lets you specify networks for wpa_supplicant declaratively:
+<programlisting>
+networking.wireless.networks = {
+  echelon = {
+    psk = "abcdefgh";
```
Comment (edolstra) [inline nixos/doc/manual/configuration/wireless.xml]: The manual should warn that these passphrases are stored world-readable in the Nix store.

Comment (copumpkin) [inline nixos/doc/manual/configuration/wireless.xml]: Or we can merge your encryption patch :smile:

**Comment by globin** [comment]
Thanks! And sorry again for not spotting that one :hear_no_evil:

**Comment by domenkozar** [comment]
I have a fix, running NixOS tests and fixing build failures on the way.

**Comment by peti** [comment]
Reverted in https://github.com/NixOS/nixpkgs/pull/12164 to fix evaluation.

**Comment by peti** [comment]
@domenkozar, sorry, I didn't see [your message](https://github.com/NixOS/nixpkgs/pull/12015#issuecomment-169090684) in time.

**Comment by domenkozar** [comment]
For the record, following fix works:

```
diff --git a/nixos/modules/profiles/installation-device.nix b/nixos/modules/profiles/installation-device.nix
index 9460327..669b697 100644
--- a/nixos/modules/profiles/installation-device.nix
+++ b/nixos/modules/profiles/installation-device.nix
@@ -51,7 +51,7 @@ with lib;
 
     # Enable wpa_supplicant, but don't start it by default.
     networking.wireless.enable = mkDefault true;
-    jobs.wpa_supplicant.startOn = mkOverride 50 "";
+    systemd.services.wpa_supplicant.wantedBy = mkOverride 50 [];
 
     # Tell the Nix evaluator to garbage collect more aggressively.
     # This is desirable in memory-constrained environments that don't
```

**Pull request merged**